include/SDL3/SDL_gpu.h

```diff
@@ -0,0 +1,2251 @@
+    SDL_GPU_COMPAREOP_ALWAYS
+} SDL_GpuCompareOp;
+
+typedef enum SDL_GpuStencilOp
+{
+    SDL_GPU_STENCILOP_KEEP,
+    SDL_GPU_STENCILOP_ZERO,
+    SDL_GPU_STENCILOP_REPLACE,
+    SDL_GPU_STENCILOP_INCREMENT_AND_CLAMP,
+    SDL_GPU_STENCILOP_DECREMENT_AND_CLAMP,
+    SDL_GPU_STENCILOP_INVERT,
+    SDL_GPU_STENCILOP_INCREMENT_AND_WRAP,
+    SDL_GPU_STENCILOP_DECREMENT_AND_WRAP
+} SDL_GpuStencilOp;
+
+typedef enum SDL_GpuBlendOp
+{
+    SDL_GPU_BLENDOP_ADD,
+    SDL_GPU_BLENDOP_SUBTRACT,
+    SDL_GPU_BLENDOP_REVERSE_SUBTRACT,
+    SDL_GPU_BLENDOP_MIN,
+    SDL_GPU_BLENDOP_MAX
+} SDL_GpuBlendOp;
+
+typedef enum SDL_GpuBlendFactor
+{
+    SDL_GPU_BLENDFACTOR_ZERO,
+    SDL_GPU_BLENDFACTOR_ONE,
+    SDL_GPU_BLENDFACTOR_SRC_COLOR,
+    SDL_GPU_BLENDFACTOR_ONE_MINUS_SRC_COLOR,
+    SDL_GPU_BLENDFACTOR_DST_COLOR,
+    SDL_GPU_BLENDFACTOR_ONE_MINUS_DST_COLOR,
+    SDL_GPU_BLENDFACTOR_SRC_ALPHA,
+    SDL_GPU_BLENDFACTOR_ONE_MINUS_SRC_ALPHA,
+    SDL_GPU_BLENDFACTOR_DST_ALPHA,
+    SDL_GPU_BLENDFACTOR_ONE_MINUS_DST_ALPHA,
+    SDL_GPU_BLENDFACTOR_CONSTANT_COLOR,
+    SDL_GPU_BLENDFACTOR_ONE_MINUS_CONSTANT_COLOR,
+    SDL_GPU_BLENDFACTOR_SRC_ALPHA_SATURATE
+} SDL_GpuBlendFactor;
+
+typedef enum SDL_GpuColorComponentFlagBits
+{
+    SDL_GPU_COLORCOMPONENT_R_BIT = 0x00000001,
+    SDL_GPU_COLORCOMPONENT_G_BIT = 0x00000002,
+    SDL_GPU_COLORCOMPONENT_B_BIT = 0x00000004,
+    SDL_GPU_COLORCOMPONENT_A_BIT = 0x00000008
+} SDL_GpuColorComponentFlagBits;
+
+typedef Uint32 SDL_GpuColorComponentFlags;
+
+typedef enum SDL_GpuFilter
+{
+    SDL_GPU_FILTER_NEAREST,
+    SDL_GPU_FILTER_LINEAR
+} SDL_GpuFilter;
+
+typedef enum SDL_GpuSamplerMipmapMode
+{
+    SDL_GPU_SAMPLERMIPMAPMODE_NEAREST,
+    SDL_GPU_SAMPLERMIPMAPMODE_LINEAR
+} SDL_GpuSamplerMipmapMode;
+
+typedef enum SDL_GpuSamplerAddressMode
+{
+    SDL_GPU_SAMPLERADDRESSMODE_REPEAT,
+    SDL_GPU_SAMPLERADDRESSMODE_MIRRORED_REPEAT,
+    SDL_GPU_SAMPLERADDRESSMODE_CLAMP_TO_EDGE
+} SDL_GpuSamplerAddressMode;
+
+/*
+ * VSYNC:
+ *   Waits for vblank before presenting.
+ *   If there is a pending image to present, the new image is enqueued for presentation.
+ *   Disallows tearing at the cost of visual latency.
+ *   When using this present mode, AcquireSwapchainTexture will block if too many frames are in flight.
+ * IMMEDIATE:
+ *   Immediately presents.
+ *   Lowest latency option, but tearing may occur.
+ *   When using this mode, AcquireSwapchainTexture will return NULL if too many frames are in flight.
+ * MAILBOX:
+ *   Waits for vblank before presenting. No tearing is possible.
+ *   If there is a pending image to present, the pending image is replaced by the new image.
+ *   Similar to VSYNC, but with reduced visual latency.
+ *   When using this mode, AcquireSwapchainTexture will return NULL if too many frames are in flight.
+ */
+typedef enum SDL_GpuPresentMode
+{
+    SDL_GPU_PRESENTMODE_VSYNC,
+    SDL_GPU_PRESENTMODE_IMMEDIATE,
+    SDL_GPU_PRESENTMODE_MAILBOX
+} SDL_GpuPresentMode;
```

Comment:
What about adaptive (late tear) mode? See `SDL_RENDERER_VSYNC_ADAPTIVE` for details.

Reply:
I think we ended up cutting this because it only worked on X11 and Windows, and only with Vulkan - so long-term it would have been just Windows Vulkan. We could be convinced otherwise if SDL_Render's made it work elsewhere?

Reply:
You're right, this ended up being Vulkan only.

Reply:
Actually, I think this works with OpenGL as well.

Reply:
WGL_EXT_swap_control_tear and GLX_EXT_swap_control_tear should work, I don't think the EGL variant ever got drafted - so it's actually in a pretty similar situation; I wonder if the WGL extension still works on the latest Windows anymore.

Reply:
(But the GPU API doesn't (and likely will never) support OpenGL, so it's moot.)